File: read_rays.cpp
#include "rays_generator.cpp"
#include <iostream>
#include <fstream>
#include <ctime>

int main() {
    std::string test_path = "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold";
    
    fs::path msgpack_file = fs::path(test_path) / ".rays" / "files.msgpack";
    
    if (!fs::exists(msgpack_file)) {
        std::cout << "Error: " << msgpack_file << " does not exist!" << std::endl;
        return 1;
    }
    
    std::ifstream ifs(msgpack_file, std::ios::binary);
    
    std::string buffer_str((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
    
    msgpack::object_handle oh = msgpack::unpack(buffer_str.data(), buffer_str.size());
    
    msgpack::object obj = oh.get();
    
    std::vector<FileRecord> file_registry;
    obj.convert(file_registry);
    
    std::cout << "==================================================" << std::endl;
    std::cout << "Total files found: " << file_registry.size() << std::endl;
    std::cout << "==================================================" << std::endl << std::endl;
    
    for (const auto& record : file_registry) {
        std::cout << "File: " << record.relative_path << std::endl;
        std::cout << "  Type: " << record.file_type << std::endl;
        std::cout << "  Language: " << record.language << std::endl;
        std::cout << "  State: " << record.existence_state << std::endl;
        std::cout << "  Hash: " << record.stable_id << std::endl;
        std::cout << "  Size: " << record.file_size << " bytes" << std::endl;
        
        std::time_t t = record.last_modified;
        std::cout << "  Modified: " << std::ctime(&t);
        
        std::cout << "--------------------------------------------------" << std::endl;
    }
    
    std::cout << std::endl << "Total: " << file_registry.size() << " files" << std::endl;
    
    return 0;
}

File: read_rays.py
"""
read_rays.py  —  Python equivalent of read_rays.cpp
Reads and prints .rays/files.msgpack
Usage: python read_rays.py <project_root>
"""
import sys
from pathlib import Path
from datetime import datetime
import msgpack

def main():
    test_path = sys.argv[1] if len(sys.argv) > 1 else \
        "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold"

    msgpack_file = Path(test_path) / ".rays" / "files.msgpack"

    if not msgpack_file.exists():
        print(f"Error: {msgpack_file} does not exist!")
        print("Run 'python rays_generator.py' first to generate it.")
        sys.exit(1)

    with open(msgpack_file, "rb") as f:
        file_registry = msgpack.unpack(f, raw=False)

    print("==================================================")
    print(f"Total files found: {len(file_registry)}")
    print("==================================================\n")

    for record in file_registry:
        print(f"File: {record['relative_path']}")
        print(f"  Type:     {record['file_type']}")
        print(f"  Language: {record['language']}")
        print(f"  State:    {record['existence_state']}")
        print(f"  Hash:     {record['stable_id']}")
        print(f"  Size:     {record['file_size']} bytes")
        ts = record['last_modified']
        print(f"  Modified: {datetime.fromtimestamp(ts).strftime('%a %b %d %H:%M:%S %Y')}")
        print("--------------------------------------------------")

    print(f"\nTotal: {len(file_registry)} files")

if __name__ == "__main__":
    main()
